### PR TITLE
added a fixed control period to loop

### PR DESCRIPTION
**Pull request by jackcenter**
This uses a fixed control period to remove the drift issue brought up in Issue #644. It also has the time adjustment from PR #646. I ran some simple tests based on the attachment replacing `controller_manager/ros2_control_node.cpp` (similar to what was done in #644) while running the rrbot (`ros2 launch ros2_control_demo_bringup rrbot.launch.py`) using 10 and 1,000 Hz control loops for 10 seconds.  

At 1000Hz the current implementation and PR #646 drift almost a full second over the 10-second experiment. This proposed implementation doesn't show any signs of drift over the same test. The data from my quick experiment is [here](https://docs.google.com/spreadsheets/d/171hXZe9Kps0NN-Hutd5xFktWDz_Cl8S0VuuRP88dyvw/edit?usp=sharing).

I wasn't sure if the period time would need to be updated in the loop or not (I didn't think so), but please let me know if you think it should be moved.

(edit: added test file attachment)
[ros2_control_node.cpp.txt](https://github.com/ros-controls/ros2_control/files/8081195/ros2_control_node.cpp.txt)

### CODE REVIEW

**Review by bmagyar — APPROVED** [review]
Thanks guys for opening #644  and the data in this PR.

**Comment by bmagyar** [comment]
And it seems this should be backported to Foxy and Galactic too

**Comment by gavanderhoorn** [comment]
Has moving to a timer been discussed?

I'm not sure whether the current state of the executors is such that they can be used for reliable timing, but there are other alternatives to setup timer callbacks (ie: non-ROS).

Using timers / callbacks would:

 1. be quite similar to how you'd do this in other control frameworks
 2. avoid cumulative drift
 3. if backed by a hw-timer: be potentially *very* accurate
 4. be easier to read / grok
 5. allow detecting overruns easier (either the OS or the RTL typically has support for that)

any type of `sleep(..)` always has the potential to drift in my experience.

**Comment by bmagyar** [comment]
We haven't discussed this option yet, no that I'm aware of at least.
However the secondary point with this node was to give a good example for people in case they need to implement their own loop for a reason. I think we could have a second node (we can debate which should be the default one ;) ) that uses timers.

**Comment by gavanderhoorn** [comment]
I should probably move the comment to #644.

I do believe timers should be the main example, but that's a different discussion.

File: controller_manager/src/ros2_control_node.cpp [hunk 1]
```diff
@@ -41,21 +41,29 @@ int main(int argc, char ** argv)
...
+    std::this_thread::sleep_for(std::chrono::nanoseconds(
+      (end_loop - cm->now()).nanoseconds()));  // wait until we hit the end of the period
+    end_loop += period;
+
     while (rclcpp::ok())
     {
-      rclcpp::Time begin_last = begin;
       begin = cm->now();
       cm->read();
       cm->update(begin, begin - begin_last);
```
Comment (destogl) [inline controller_manager/src/ros2_control_node.cpp:59]:
```suggestion
      // wait until we hit the end of the period
      end_period += period;
      std::this_thread::sleep_for(std::chrono::nanoseconds(
        (end_period - cm->now()).nanoseconds()));

      // execute "real-time" update loop
      cm->read();
      current_time = cm->now();
      cm->update(current_time, current_time - previous_time);
```

File: controller_manager/src/ros2_control_node.cpp [hunk 2]
```diff
@@ -41,21 +41,29 @@ int main(int argc, char ** argv)
...
+      begin_last = begin;
       cm->write();
-      rclcpp::Time end = cm->now();
-      std::this_thread::sleep_for(std::max(
-        std::chrono::nanoseconds(0),
-        std::chrono::nanoseconds(1000000000 / cm->get_update_rate()) -
-          std::chrono::nanoseconds(end.nanoseconds() - begin.nanoseconds())));
+      end = cm->now();
```
Comment (destogl) [inline controller_manager/src/ros2_control_node.cpp:62]:
```suggestion
```

File: controller_manager/src/ros2_control_node.cpp [hunk 3]
```diff
@@ -41,21 +41,29 @@ int main(int argc, char ** argv)
...
+    rclcpp::Time end = begin;
 
     // Use nanoseconds to avoid chrono's rounding
-    std::this_thread::sleep_for(std::chrono::nanoseconds(1000000000 / cm->get_update_rate()));
+    rclcpp::Duration period(std::chrono::nanoseconds(1000000000 / cm->get_update_rate()));
+    rclcpp::Time end_loop = begin + period;
+
+    std::this_thread::sleep_for(std::chrono::nanoseconds(
+      (end_loop - cm->now()).nanoseconds()));  // wait until we hit the end of the period
+    end_loop += period;
```
Comment (destogl) [inline controller_manager/src/ros2_control_node.cpp:53]:
```suggestion
```

File: controller_manager/src/ros2_control_node.cpp [hunk 4]
```diff
@@ -41,21 +41,29 @@ int main(int argc, char ** argv)
...
-      rclcpp::Time end = cm->now();
-      std::this_thread::sleep_for(std::max(
-        std::chrono::nanoseconds(0),
-        std::chrono::nanoseconds(1000000000 / cm->get_update_rate()) -
-          std::chrono::nanoseconds(end.nanoseconds() - begin.nanoseconds())));
+      end = cm->now();
+
+      std::this_thread::sleep_for(std::chrono::nanoseconds(
+        (end_loop - cm->now()).nanoseconds()));  // wait until we hit the end of the period
+      end_loop += period;
```
Comment (destogl) [inline controller_manager/src/ros2_control_node.cpp:66]:
```suggestion
```

**Review by destogl — CHANGES_REQUESTED** [review]
In light of clearer naming of variable as proposed in #645 and trying to make the code cleaner by reducing some lines, I propose the following updates. 

@jackcenter can you test this one more time? My proposal should not change the behavior in any way. Sorry for additional effort, but I hope to solve this once for all :)

@bmagyar you will have to review this again.

**Comment by jackcenter** [comment]
@destogl I made the requested changes and it is much easier to read now, thanks! I ran the tests again and there was no loss in performance or apparent drift.

**Review by AndyZe — APPROVED** [review]

**Review by aprotyas — APPROVED** [review]

**Review by destogl — APPROVED** [review]
Thanks for the great contribution and detailed analysis!! :rocket:

**Review by bmagyar — APPROVED** [review]

**Pull request merged**